internal/controller/common.go

```diff
@@ -3,32 +3,258 @@ package controller
+	return &common{
+		addonGetterFactory: addonGetterFactory,
+
+		addonStorage: addonStorage,
+		chartStorage: chartStorage,
+
+		brokerFacade: brokerFacade,
+		brokerSyncer: brokerSyncer,
+
+		docsProvider: docsProvider,
+		protection:   protection{},
+
+		namespace:    internal.ClusterWide,
+		commonClient: NewCommonClient(client, log),
+
+		dstPath: dstPath,
+		log:     log,
+	}
+}
+
+// SetWorkingNamespace sets services's working namespace. It should only be used by namespace-scoped controller.
+func (c *common) SetWorkingNamespace(namespace string) {
+	c.namespace = internal.Namespace(namespace)
+	for _, svc := range []NamespacedService{
+		c.brokerSyncer, c.brokerFacade, c.docsProvider, c.commonClient,
+	} {
+		svc.SetNamespace(namespace)
+	}
+}
+
+// Reconcile handles configuration state change
+func (c *common) Reconcile(addon *internal.CommonAddon, trace string) (reconcile.Result, error) {
+	if addon.Meta.DeletionTimestamp != nil {
+		c.log.Infof("Start delete %s", trace)
+
+		if err := c.OnDelete(addon); err != nil {
+			c.log.Errorf("while deleting %s process: %v", trace, err)
+			return reconcile.Result{RequeueAfter: time.Second * 15}, errors.Wrapf(err, "while deleting %s", trace)
+		}
+		c.log.Infof("Delete %s process completed", trace)
+		return reconcile.Result{}, nil
+	}
+
+	if addon.Status.ObservedGeneration == 0 {
+		c.log.Infof("Start add %s process", trace)
+
+		if err := c.PrepareForProcessing(addon); err != nil {
+			c.log.Errorf("while preparing %s for processing: %v", trace, err)
+			return reconcile.Result{}, errors.Wrapf(err, "while preparing %s for processing", trace)
+		}
+		if err := c.OnAdd(addon, addon.Status); err != nil {
+			c.log.Errorf("while adding %s process: %v", trace, err)
+			return reconcile.Result{}, errors.Wrapf(err, "while creating %s", trace)
+		}
+		c.log.Infof("Add %s process completed", trace)
+
+	} else if addon.Meta.Generation > addon.Status.ObservedGeneration {
+		c.log.Infof("Start update %s process", trace)
+
+		lastStatus := addon.Status
+		addon.Status = v1alpha1.CommonAddonsConfigurationStatus{}
+
+		if err := c.OnAdd(addon, lastStatus); err != nil {
+			c.log.Errorf("while updating %s process: %v", trace, err)
+			return reconcile.Result{}, errors.Wrapf(err, "while updating %s", trace)
+		}
+		c.log.Infof("Update %s process completed", trace)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// PrepareForProcessing prepares ClusterAddonsConfiguration or AddonsConfiguration if namespace is set
+func (c *common) PrepareForProcessing(addon *internal.CommonAddon) error {
+	err := c.addFinalizer(addon)
+	if err != nil {
+		return errors.Wrap(err, "while adding finalizer")
+	}
+	addon.Status.Phase = v1alpha1.AddonsConfigurationPending
+	err = c.updateAddonStatus(addon)
+	if err != nil {
+		return errors.Wrap(err, "while updating status")
+	}
+
+	return nil
```

Comment:
Is it necessary to update `addon` in two steps? First after adding finalizer, second after change status to "Pending"? Can we do this in one step?
- add finalizer if necessary
- change(add) status
- update `Addon`

Reply:
we cannot do like that, the status is a subresource and we must update it separately using the `Status()` client